File: sap-android/app/src/main/java/edu/csuft/sap/ui/theme/Color.kt
package edu.csuft.sap.ui.theme

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.luminance

// 强调色（辅色）已迁至 ThemeState：默认天蓝、可在设置里换。此处仅保留简约白底色与中性灰。

// 简约白 + 中性灰
val PageBg = Color(0xFFF6F7F9)
val Surface = Color(0xFFFFFFFF)
val SurfaceMuted = Color(0xFFF1F3F5)
val TextPrimary = Color(0xFF1F2329)
val TextSecondary = Color(0xFF646A73)
val TextTertiary = Color(0xFF9098A1)
val Hairline = Color(0xFFEDEFF2)
val Danger = Color(0xFFE2504A)

/** 课程色卡：每门课按名称稳定取一组（底色, 文字色）。 */
data class CourseColor(val container: Color, val onContainer: Color)

// WakeUp 风格 12 色柔和卡片（底色, 同色系深字）——无绿/青绿系；底色比纯淡再加深一档（约 84% 亮度）
val CoursePalette: List<CourseColor> = listOf(
    CourseColor(Color(0xFFC5DDF7), Color(0xFF2A6BB0)), // 0 blue
    CourseColor(Color(0xFFF8CEBB), Color(0xFFC0552C)), // 1 coral
    CourseColor(Color(0xFFCCD1F0), Color(0xFF45499C)), // 2 indigo
    CourseColor(Color(0xFFD8D0F6), Color(0xFF5A4FBE)), // 3 purple
    CourseColor(Color(0xFFF8DDA6), Color(0xFFB07A12)), // 4 amber
    CourseColor(Color(0xFFF8CADC), Color(0xFFB53E6E)), // 5 pink
    CourseColor(Color(0xFFC7D8EC), Color(0xFF3A6491)), // 6 steel blue
    CourseColor(Color(0xFFF8D6AD), Color(0xFFB16A18)), // 7 apricot
    CourseColor(Color(0xFFF8CCCC), Color(0xFFBD4B4B)), // 8 rose
    CourseColor(Color(0xFFCFD5DF), Color(0xFF4C566B)), // 9 slate
    CourseColor(Color(0xFFDDD3F4), Color(0xFF6A5BC0)), // 10 lavender
    CourseColor(Color(0xFFEFE3A4), Color(0xFF9A7B10)), // 11 gold
)

/** 名称 → 稳定色卡下标（同名课永远同色）。 */
fun colorIndexOf(name: String?): Int {
    val key = name.orEmpty()
    if (key.isEmpty()) return 0
    var h = 0
    for (c in key) h = (h * 31 + c.code) and 0x7fffffff
    return h % CoursePalette.size
}

/** 取色卡：override 优先（自建课指定色），否则按名称稳定取色。 */
fun courseColorOf(name: String?, override: Int? = null): CourseColor {
    val idx = if (override != null && override in CoursePalette.indices) override else colorIndexOf(name)
    return CoursePalette[idx]
}

/** 安全取调色板色卡。 */
fun paletteColor(index: Int): CourseColor =
    CoursePalette[((index % CoursePalette.size) + CoursePalette.size) % CoursePalette.size]

/** 由单个自定义颜色(ARGB)派生课卡色卡：选中色作底，按亮度自动配可读文字色。 */
fun customCourseColor(argb: Long): CourseColor {
    val container = Color(argb)
    val onC = if (container.luminance() > 0.5f)
        Color(container.red * 0.32f, container.green * 0.32f, container.blue * 0.32f, 1f)
    else Color.White
    return CourseColor(container, onC)
}

File: sap-weapp/utils/color.js
/**
 * 课程色卡（移植自安卓 Color.kt 的 CoursePalette）。
 * 12 色柔和卡片：{ container 底色, onContainer 文字色 }。
 */
const CoursePalette = [
  { container: '#C5DDF7', onContainer: '#2A6BB0' }, // 0 blue
  { container: '#F8CEBB', onContainer: '#C0552C' }, // 1 coral
  { container: '#CCD1F0', onContainer: '#45499C' }, // 2 indigo
  { container: '#D8D0F6', onContainer: '#5A4FBE' }, // 3 purple
  { container: '#F8DDA6', onContainer: '#B07A12' }, // 4 amber
  { container: '#F8CADC', onContainer: '#B53E6E' }, // 5 pink
  { container: '#C7D8EC', onContainer: '#3A6491' }, // 6 steel blue
  { container: '#F8D6AD', onContainer: '#B16A18' }, // 7 apricot
  { container: '#F8CCCC', onContainer: '#BD4B4B' }, // 8 rose
  { container: '#CFD5DF', onContainer: '#4C566B' }, // 9 slate
  { container: '#DDD3F4', onContainer: '#6A5BC0' }, // 10 lavender
  { container: '#EFE3A4', onContainer: '#9A7B10' }, // 11 gold
]

/** 名称 → 稳定色卡下标（同名课永远同色）。 */
function colorIndexOf(name) {
  const key = name || ''
  if (!key) return 0
  let h = 0
  for (let i = 0; i < key.length; i++) h = (h * 31 + key.charCodeAt(i)) & 0x7fffffff
  return h % CoursePalette.length
}

/** 取色卡：override 优先，否则按名称稳定取色。 */
function courseColorOf(name, override) {
  const idx = (override != null && override >= 0 && override < CoursePalette.length) ? override : colorIndexOf(name)
  return CoursePalette[idx]
}

function paletteColor(index) {
  const n = CoursePalette.length
  return CoursePalette[((index % n) + n) % n]
}

/** 由 #RRGGBB 计算相对亮度（0..1）。 */
function luminance(hex) {
  const m = /^#?([0-9a-fA-F]{6})$/.exec(hex)
  if (!m) return 0
  const v = parseInt(m[1], 16)
  const r = ((v >> 16) & 255) / 255, g = ((v >> 8) & 255) / 255, b = (v & 255) / 255
  return 0.2126 * r + 0.7152 * g + 0.0722 * b
}

/** 由单个自定义颜色派生课卡色卡：选中色作底，按亮度配可读文字色。 */
function customCourseColor(hex) {
  const onC = luminance(hex) > 0.5 ? darken(hex, 0.32) : '#FFFFFF'
  return { container: hex, onContainer: onC }
}

function darken(hex, factor) {
  const m = /^#?([0-9a-fA-F]{6})$/.exec(hex)
  if (!m) return hex
  const v = parseInt(m[1], 16)
  const r = Math.round(((v >> 16) & 255) * factor)
  const g = Math.round(((v >> 8) & 255) * factor)
  const b = Math.round((v & 255) * factor)
  return '#' + [r, g, b].map((x) => x.toString(16).padStart(2, '0')).join('')
}

module.exports = { CoursePalette, colorIndexOf, courseColorOf, paletteColor, customCourseColor, luminance }
